extensions/amp-link-rewriter/validator-amp-link-rewriter.protoascii

```diff
@@ -0,0 +1,60 @@
+#
+# Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the license.
+#
+tags: {  # amp-link-rewriter
```

Comment:
It will be great if we can enforce that `<amp-link-rewriter>` doesn't come with `<amp-skimlink>` and `<amp-smartlink>`.  How difficult it is to support that?

Reply:
Is the request that if `<amp-link-rewriter>` is on a document then there should also be an error if that document also included `<amp-skimlink>` or `<amp-smartlink>`? Currently there isn't a method of excluding other extensions but it could be looked into. @ampproject/wg-caching for interest in extensions excluding other extensions.

Reply:
That's right, because those three components can all rewrite link on the page. The ideal approach is to have the two previous solution migrate to the generic one. 
Thanks for the information. I think we are good now, I don't see an incentive to include more.

Reply:
Sounds good then. If excluding other extensions when an extension is present is useful then we may want to create an issue to request that.

Out of curiosity, what would happen if all three (`amp-link-rewriter`, `amp-skimlink` and `amp-smartlink`) were included on the page? Is there a defined order based on which is included when?

Reply:
Thank you. I created #21723. 

Right now `<amp-link-rewriter>` runs last, but we don't control which `amp-skimlink` or `amp-smartlink` runs first. 
I just realize that one can include all three components, and still have them work well because they can be configured to rewrite different links on the page. But I also don't expect that use case to exist : )

Reply:
Ah-ha! While we can't exclude specific extensions we can enforce this through another mechanism. The validator has `satisfies`, `requires`, and `excludes` for tagspecs. [validator.proto](https://github.com/ampproject/amphtml/blob/master/validator/validator.proto#L546).

`<amp-link-rewriter>` would have:

```
satisfies: "amp-link-rewriter"
excludes: "amp-skimlink"
excludes: "amp-smartlink"
```

`<amp-skimlink>` would have:

```
satisfies: "amp-skimlink"
excludes: "amp-link-rewriter"
excludes: "amp-smartlink"
```

`<amp-smartlink>` would have:

```
satisfies: "amp-smartlink"
excludes: "amp-link-rewriter"
excludes: "amp-skimlink"
```

Reply:
Ha! The magic `satisfies` `requires` and `excludes` that help with all the use cases! @parsingeye Could you please add this to `<amp-link-rewrite>`. We can change validator rules of the `<amp-skimlink>` and `<amp-smartlink>` in different PRs.

Reply:
Hi guys, we are trying to achieve this but it seems not to work I'm attaching a sample of how we are adding it in our .protoascii file:

**tags: {  # <amp-link-rewriter>
  html_format: AMP
  tag_name: "AMP-LINK-REWRITER"
  unique: true
  requires_extension: "amp-link-rewriter"
  satisfies: "amp-link-rewriter"
  excludes: "amp-skimlinks"
  excludes: "amp-smartlinks"
  child_tags: {
    mandatory_num_child_tags: 1
    first_child_tag_name_oneof: "SCRIPT"
  }
  amp_layout: {
    supported_layouts: NODISPLAY
  }
}**

Reply:
Could you please advise? cc: @honeybadgerdontcare @torch2424

Reply:
Hello!

@es7eban @parsingeye 

Oh that is odd! But in terms of help writing the protoascii, here are some good resources:

[A guide on building a simple extensions, and getting it validated](https://github.com/ampproject/amphtml/blob/master/contributing/component-validator-rules.md)

[A simple protoascii for a standalone gist embed extension](https://github.com/ampproject/amphtml/blob/master/extensions/amp-gist/validator-amp-gist.protoascii)

Let me know if these help, or if there is anything in particular about your current protoascii (or results from trying to run the protoascii) that are errors I can help with.

Thanks! 😄 

P.S Could we perhaps move the discussion to a new "Validation PR" that you can open? 🤔 

cc @Gregable while @honeybadgerdontcare is out travelling.